### PR TITLE
feat: allocating connection id on server side

### PR DESCRIPTION
**Pull request by woorui**
# Description

resolve #710 gracefully.

## Impact

The server no longer uses the ID from the client as the storage key. Instead, it now uses an automatically incremented integer as the storage key.

### CODE REVIEW

File: core/connection.go [hunk 1]
```diff
@@ -50,11 +64,16 @@ func newConnection(
 	}
 }
 
-// ID returns the connection ID.
-func (c *Connection) ID() string {
+// ID returns the increment ID.
```
Comment (fanweixiao) [inline core/connection.go:67]:
```suggestion
// ID returns the connection ID.
```

this `ID` property always represents the connection ID, developers do not need to understand it's auto-incremented and focus on its use.

File: core/connector.go [hunk 2]
```diff
@@ -32,7 +33,7 @@ func NewConnector(ctx context.Context) *Connector {
 // Store stores Connection to Connector,
 // If the connID is the same twice, the new connection will replace the old connection.
```
Comment (fanweixiao) [inline core/connector.go:34]:
```suggestion
// The newer connection will replaces the older one.
```

File: core/connector.go [hunk 3]
```diff
@@ -32,7 +33,7 @@ func NewConnector(ctx context.Context) *Connector {
...
 // If Connector be closed, The function will return ErrConnectorClosed.
```
Comment (fanweixiao) [inline core/connector.go:35]:
```suggestion
// If a Connector is closed, the function returns ErrConnectorClosed.
```

File: core/connector.go [hunk 4]
```diff
@@ -47,7 +48,7 @@ func (c *Connector) Store(connID string, conn *Connection) error {
 // Remove removes the connection with the specified connID.
 // If the Connector does not have a connection with the given connID, no action is taken.
 // If Connector be closed, The function will return ErrConnectorClosed.
```
Comment (fanweixiao) [inline core/connector.go:50]:
```suggestion
// If a Connector is closed, the function returns ErrConnectorClosed.
```

File: core/connector.go [hunk 5]
```diff
@@ -59,17 +60,17 @@ func (c *Connector) Remove(connID string) error {
 	return nil
 }
 
-// Get retrieves the Connection with the specified connID.
-// If the Connector does not have a connection with the given connID, return nil and false.
+// Get retrieves the Connection with the specified id.
+// If the Connector does not have a connection with the given id, return nil and false.
 // If Connector be closed, The function will return ErrConnectorClosed.
```
Comment (fanweixiao) [inline core/connector.go:65]:
```suggestion
// If a Connector is closed, the function returns ErrConnectorClosed.
```

**Review by fanweixiao — CHANGES_REQUESTED** [review]

File: core/connection.go [hunk 6]
```diff
@@ -1,17 +1,28 @@
...
+func incrID() uint64 {
+	return atomic.AddUint64(&increment, 1)
+}
+
 // ConnectionInfo holds the information of connection.
 type ConnectionInfo interface {
+	// ID is the ID generated by server.
+	ID() uint64
+	// ClientID represents a client ID, the ClientID generated by client.
+	ClientID() string
```
Comment (venjiang) [inline core/connection.go:23]: Can the client ID be deleted?

Comment (woorui) [reply core/connection.go:23]: I think we can keep it for now. We can trace the client by the clientID. 

And the clientID is from handshakeFrame.

**Comment by fanweixiao** [comment]
@wujunzhuo please review this PR.

**Review by fanweixiao — APPROVED** [review]

**Pull request merged**